sim/src/main/scala/spinal/sim/VerilatorBackend.scala

```diff
@@ -460,15 +483,59 @@ JNIEXPORT void API JNICALL ${jniPrefix}disableWave_1${uniqueId}
        | --Mdir ${workspaceName}
        | --top-module ${config.toplevelName}
        | $rtlIncludeDirsArgs
-       | -cc ${config.rtlSourcesPaths.filter(e => e.endsWith(".v") || 
-                                                  e.endsWith(".sv") || 
-                                                  e.endsWith(".h"))
-                                     .map(new File(_).getAbsolutePath)
-                                     .map('"' + _.replace("\\","/") + '"')
-                                     .mkString(" ")}
+       | -cc ${verilatorInputFiles.map('"' + _.replace("\\","/") + '"').mkString(" ")}
        | --exe $workspaceName/$wrapperCppName
        | ${config.simulatorFlags.mkString(" ")}""".stripMargin.replace("\n", "")
 
+
+    if (cacheEnabled) {
+      // calculate hash of verilator version+options and source file contents
+
+      val md = MessageDigest.getInstance("SHA-1")
+      md.update(verilatorScript.getBytes())
+
+      val verilatorVersionProcess = Process(Seq(verilatorBinFilename, "--version"), new File(workspacePath))
+      val verilatorVersion = verilatorVersionProcess.lineStream.mkString("\n")    // blocks and throws an exception if exit status != 0
+      md.update(verilatorVersion.getBytes())
+
+      verilatorInputFiles.foreach { filename =>
+        val bis = new BufferedInputStream(new FileInputStream(filename))
+        val buf = new Array[Byte](1024)
+
+        Iterator.continually(bis.read(buf, 0, buf.length))
+          .takeWhile(_ >= 0)
+          .foreach(md.update(buf, 0, _))
+
+        bis.close()
+      }
+
+      val hash = md.digest().map(x => (x & 0xFF).toHexString.padTo(2, '0')).mkString("")
+
+      val previousHashFilename = s"${cachePath}/filesHash.txt"
+      val previousHash = (new File(previousHashFilename)).exists() match {
+        case true => Source.fromFile(previousHashFilename).getLines().next().trim()
+        case false => null
+      }
+
+      // skip verilator compilation if nothing changed (hashes equal)
+      if (hash == previousHash) {
+        val workspaceDir = new File(s"${workspacePath}/${workspaceName}")
+        val cacheDir = new File(s"${cachePath}/${workspaceName}")
+
+        if (cacheDir.exists()) {
+          FileUtils.deleteQuietly(workspaceDir)
+          FileUtils.moveDirectory(cacheDir, workspaceDir)
```

Comment:
moveDirectory, should it be copyDirectory ?

Reply:
Yes you're right